Dedicated InfluxQL function reference, sample data, feature tracking (#4934)
influxdata/docs-v2: content/influxdb/cloud-dedicated/reference/influxql/_index.md

unquoted_identifier = ( letter ) { letter | digit } .
quoted_identifier   = `"` unicode_char { unicode_char } `"` .
```

#### Examples

```
cpu
_cpu_stats
"1h"
"anything really"
"1_Crazy-1337.identifier>NAME👍"
```

### Keywords

```
ALL           ALTER         ANY           AS            ASC           BEGIN
BY            CREATE        CONTINUOUS    DATABASE      DATABASES     DEFAULT
DELETE        DESC          DESTINATIONS  DIAGNOSTICS   DISTINCT      DROP
DURATION      END           EVERY         EXPLAIN       FIELD         FOR
FROM          GRANT         GRANTS        GROUP         GROUPS        IN
INF           INSERT        INTO          KEY           KEYS          KILL
LIMIT         SHOW          MEASUREMENT   MEASUREMENTS  NAME          OFFSET
ON            ORDER         PASSWORD      POLICY        POLICIES      PRIVILEGES
QUERIES       QUERY         READ          REPLICATION   RESAMPLE      RETENTION
REVOKE        SELECT        SERIES        SET           SHARD         SHARDS
SLIMIT        SOFFSET       STATS         SUBSCRIPTION  SUBSCRIPTIONS TAG
TO            USER          USERS         VALUES        WHERE         WITH
WRITE
```

> If you use an InfluxQL keywords as an
> [identifier](/influxdb/v2.7/reference/glossary/#identifier) you will need to
> double quote that identifier in every query.

Reviewer: ```suggestion
If you use an InfluxQL keyword as an
[identifier](/influxdb/cloud-dedicated/reference/glossary/#identifier),
double-quote the identifier in every query.
```

> If you use an InfluxQL keywords as an
> [identifier](/influxdb/v2.7/reference/glossary/#identifier) you will need to
> double quote that identifier in every query.

Reviewer: ```suggestion
If you use an InfluxQL keyword as an
[identifier](/influxdb/cloud-dedicated/reference/glossary/#identifier),
double-quote the identifier in every query.
```

```
float_lit           = int_lit "." int_lit .
```

#### Strings

String literals must be surrounded by single quotes.
Strings may contain `'` characters as long as they are escaped (i.e., `\'`).

```
string_lit          = `'` { unicode_char } `'` .
```

#### Durations

Duration literals specify a length of time.
An integer literal followed immediately (with no spaces) by a duration unit listed below is interpreted as a duration literal.
Durations can be specified with mixed units.

##### Duration units

| Units  | Meaning                                 |
| ------ | --------------------------------------- |
| ns     | nanoseconds (1 billionth of a second)   |
| u or µ | microseconds (1 millionth of a second)  |
| ms     | milliseconds (1 thousandth of a second) |
| s      | second                                  |
| m      | minute                                  |
| h      | hour                                    |
| d      | day                                     |
| w      | week                                    |

```
duration_lit        = int_lit duration_unit .
duration_unit       = "ns" | "u" | "µ" | "ms" | "s" | "m" | "h" | "d" | "w" .
```

#### Dates & Times

> The date and time literal format is not specified in EBNF like the rest of this document.
> It is specified using Go's date / time parsing format, which is a reference date written in the format required by InfluxQL.

Reviewer: ```suggestion
Unlike other notations used in InfluxQL, the date and time literal format isn't specified by EBNF.
InfluxQL date and time is specified using Go's time parsing format and [reference date](https://pkg.go.dev/time#pkg-constants) written in the format required by InfluxQL.
```

The reference date time is:

InfluxQL reference date time: January 2nd, 2006 at 3:04:05 PM

```
time_lit            = "2006-01-02 15:04:05.999999" | "2006-01-02" .
```

#### Booleans

```
bool_lit            = TRUE | FALSE .
```

#### Regular Expressions

```
regex_lit           = "/" { unicode_char } "/" .
```

**Comparators:**
`=~` matches against
`!~` doesn't match against

{{% note %}}
**NOTE:** InfluxQL supports using regular expressions when specifying:

> * [field keys](/influxdb/v2.7/reference/glossary/#field-key) and [tag keys](/influxdb/v2.7/reference/glossary/#tag-key) in the [`SELECT` clause](/influxdb/v2.7/query-data/influxql/explore-data/select/)
> * [measurements](/influxdb/v2.7/reference/glossary/#measurement) in the [`FROM` clause](/influxdb/v2.7/query-data/influxql/explore-data/select/#from-clause)
> * [tag values](/influxdb/v2.7/reference/glossary/#tag-value) and string [field values](/influxdb/v2.7/reference/glossary/#field-value) in the [`WHERE` clause](/influxdb/v2.7/query-data/influxql/explore-data/where/)
> * [tag keys](/influxdb/v2.7/reference/glossary/#tag-key) in the [`GROUP BY` clause](/influxdb/v2.7/query-data/influxql/explore-data/group-by/)

Reviewer: ```suggestion
* [field keys](/influxdb/cloud-dedicated/reference/glossary/#field-key) and [tag keys](/influxdb/cloud-dedicated/reference/glossary/#tag-key) in the [`SELECT` clause](/influxdb/cloud-dedicated/query-data/influxql/explore-data/select/)
* [measurements](/influxdb/cloud-dedicated/reference/glossary/#measurement) in the [`FROM` clause](/influxdb/cloud-dedicated/query-data/influxql/explore-data/select/#from-clause)
* [tag values](/influxdb/cloud-dedicated/reference/glossary/#tag-value) and string [field values](/influxdb/cloud-dedicated/reference/glossary/#field-value) in the [`WHERE` clause](/influxdb/cloud-dedicated/query-data/influxql/explore-data/where/)
* [tag keys](/influxdb/cloud-dedicated/reference/glossary/#tag-key) in the [`GROUP BY` clause](/influxdb/cloud-dedicated/query-data/influxql/explore-data/group-by/)
```

> * [field keys](/influxdb/v2.7/reference/glossary/#field-key) and [tag keys](/influxdb/v2.7/reference/glossary/#tag-key) in the [`SELECT` clause](/influxdb/v2.7/query-data/influxql/explore-data/select/)
> * [measurements](/influxdb/v2.7/reference/glossary/#measurement) in the [`FROM` clause](/influxdb/v2.7/query-data/influxql/explore-data/select/#from-clause)
> * [tag values](/influxdb/v2.7/reference/glossary/#tag-value) and string [field values](/influxdb/v2.7/reference/glossary/#field-value) in the [`WHERE` clause](/influxdb/v2.7/query-data/influxql/explore-data/where/)
> * [tag keys](/influxdb/v2.7/reference/glossary/#tag-key) in the [`GROUP BY` clause](/influxdb/v2.7/query-data/influxql/explore-data/group-by/)

Reviewer: ```suggestion
* [field keys](/influxdb/cloud-dedicated/reference/glossary/#field-key) and [tag keys](/influxdb/cloud-dedicated/reference/glossary/#tag-key) in the [`SELECT` clause](/influxdb/cloud-dedicated/query-data/influxql/explore-data/select/)
* [measurements](/influxdb/cloud-dedicated/reference/glossary/#measurement) in the [`FROM` clause](/influxdb/cloud-dedicated/query-data/influxql/explore-data/select/#from-clause)
* [tag values](/influxdb/cloud-dedicated/reference/glossary/#tag-value) and string [field values](/influxdb/cloud-dedicated/reference/glossary/#field-value) in the [`WHERE` clause](/influxdb/cloud-dedicated/query-data/influxql/explore-data/where/)
* [tag keys](/influxdb/cloud-dedicated/reference/glossary/#tag-key) in the [`GROUP BY` clause](/influxdb/cloud-dedicated/query-data/influxql/explore-data/group-by/)
```

> Currently, InfluxQL does not support using regular expressions to match non-string field values in the `WHERE` clause, [databases](/influxdb/v2.7/reference/glossary/#database), and [retention polices](/influxdb/v2.7/reference/glossary/#retention-policy-rp).

Reviewer: ```suggestion
Currently, InfluxQL doesn't support using regular expressions to match non-string field values in the `WHERE` clause, [databases](/influxdb/cloud-dedicated/reference/glossary/#database), and [retention polices](/influxdb/cloud-dedicated/reference/glossary/#retention-policy-rp).
```

{{% /note %}}

## Queries

> A query is composed of one or more statements separated by a semicolon.

Reviewer: ```suggestion
A query is composed of one or more statements separated by a semicolon (`;`).
```

```
query               = statement { ";" statement } .

statement           = explain_stmt |
                      explain_analyze_stmt |
                      select_stmt |
                      show_databases_stmt |
                      show_field_key_cardinality_stmt |
                      show_field_keys_stmt |
                      show_measurement_exact_cardinality_stmt |
                      show_measurements_stmt |
                      show_series_exact_cardinality_stmt |
                      show_series_stmt |
                      show_tag_key_cardinality_stmt |
                      show_tag_key_exact_cardinality_stmt |
                      show_tag_keys_stmt |
                      show_tag_values_with_key = stmt |
                      show_tag_values_cardinality_stmt .
```

## Statements

### EXPLAIN

Parses and plans the query, and then prints a summary of estimated costs.

Many SQL engines use the `EXPLAIN` statement to show join order, join algorithms, and predicate and expression pushdown.

> Since InfluxQL does not support joins, the cost of a InfluxQL query is typically a function of the total series accessed, the number of iterator accesses to a TSM file, and the number of TSM blocks that need to be scanned.

Reviewer: ```suggestion
Since InfluxQL doesn't support joins, the cost of an InfluxQL query is typically a function of the total series accessed, the number of iterator accesses to a TSM file, and the number of TSM blocks that need to be scanned.
```

> The elements of `EXPLAIN` query plan include:

Reviewer: ```suggestion
A query plan generated by `EXPLAIN` contains the following elements:
```

- expression
- auxillary fields
- number of shards
- number of series
- cached values
- number of files
- number of blocks
- size of blocks

```
explain_stmt = "EXPLAIN" select_stmt .
```

#### Example

```sql
> explain select sum(pointReq) from "_internal"."monitor"."write" group by hostname;
> QUERY PLAN
------
EXPRESSION: sum(pointReq::integer)
NUMBER OF SHARDS: 2
NUMBER OF SERIES: 2
CACHED VALUES: 110
NUMBER OF FILES: 1
NUMBER OF BLOCKS: 1
SIZE OF BLOCKS: 931
```

### EXPLAIN ANALYZE

> Executes the specified SELECT statement and returns data on the query performance and storage during runtime, visualized as a tree. Use this statement to analyze query performance and storage, including [execution time](#execution-time) and [planning time](#planning-time), and the [iterator type](#iterator-type) and [cursor type](#cursor-type).

Reviewer: ```suggestion
Executes the specified `SELECT` statement and returns data about the query performance and storage during runtime, visualized as a tree. Use this statement to analyze query performance and storage, including [execution time](#execution-time) and [planning time](#planning-time), and the [iterator type](#iterator-type) and [cursor type](#cursor-type).
```

> For example, executing the following statement:

Reviewer: ```suggestion
For example, if you execute the following statement:
```

```sql
> explain analyze select mean(usage_steal) from cpu where time >= '2018-02-22T00:00:00Z' and time < '2018-02-22T12:00:00Z'
```

> May produce an output similar to the following:

Reviewer: ```suggestion
The output is similar to the following:
```

```sql
EXPLAIN ANALYZE
---------------
.
└── select
    ├── execution_time: 2.25823ms
    ├── planning_time: 18.381616ms
    ├── total_time: 20.639846ms
    └── field_iterators
        ├── labels
        │   └── statement: SELECT mean(usage_steal::float) FROM telegraf."default".cpu
        └── expression
            ├── labels
            │   └── expr: mean(usage_steal::float)
            └── create_iterator
                ├── labels
                │   ├── measurement: cpu
                │   └── shard_id: 608
                ├── cursors_ref: 779
                ├── cursors_aux: 0
                ├── cursors_cond: 0
                ├── float_blocks_decoded: 431
                ├── float_blocks_size_bytes: 1003552
                ├── integer_blocks_decoded: 0
                ├── integer_blocks_size_bytes: 0
                ├── unsigned_blocks_decoded: 0
                ├── unsigned_blocks_size_bytes: 0
                ├── string_blocks_decoded: 0
                ├── string_blocks_size_bytes: 0
                ├── boolean_blocks_decoded: 0
                ├── boolean_blocks_size_bytes: 0
                └── planning_time: 14.805277ms```
```

> > Note: EXPLAIN ANALYZE ignores query output, so the cost of serialization to JSON or CSV is not accounted for.

Reviewer: ```suggestion
> Note: EXPLAIN ANALYZE ignores query output, so the cost of serialization to JSON or CSV isn't accounted for.
```

> > Note: EXPLAIN ANALYZE ignores query output, so the cost of serialization to JSON or CSV is not accounted for.

Reviewer: ```suggestion
> Note: EXPLAIN ANALYZE ignores query output, so the cost of serialization to JSON or CSV isn't accounted for.
```

##### execution_time

Shows the amount of time the query took to execute, including reading the time series data, performing operations as data flows through iterators, and draining processed data from iterators. Execution time doesn't include the time taken to serialize the output into JSON or other formats.

##### planning_time

Shows the amount of time the query took to plan.

> Planning a query in InfluxDB requires a number of steps. Depending on the complexity of the query, planning can require more work and consume more CPU and memory resources than the executing the query. For example, the number of series keys required to execute a query affects how quickly the query is planned and the required memory.

Reviewer: ```suggestion
Planning a query in InfluxDB requires a number of steps. Depending on the complexity of the query, planning can require more work and consume more CPU and memory resources than executing the query. For example, the number of series keys required to execute a query affects how quickly the query is planned and how much memory the planning requires.
```

> First, InfluxDB determines the effective time range of the query and selects the shards to access (in InfluxDB Enterprise, shards may be on remote nodes).

Reviewer: ```suggestion
First, InfluxDB determines the effective time range of the query and selects the shards to access.
```

##### block types

EXPLAIN ANALYZE separates storage block types, and reports the total number of blocks decoded and their size (in bytes) on disk. The following block types are supported:

| `float`    | 64-bit IEEE-754 floating-point number |
| `integer`  | 64-bit signed integer                 |
| `unsigned` | 64-bit unsigned integer               |
| `boolean`  | 1-bit, LSB encoded                    |
| `string`   | UTF-8 string                          |

For more information about storage blocks, see [TSM files](/influxdb/v2.7/reference/internals/storage-engine/#time-structured-merge-tree-tsm).

### SELECT

```
select_stmt = "SELECT" fields from_clause [ where_clause ]
              [ group_by_clause ] [ order_by_clause ] [ limit_clause ]
              [ offset_clause ] [ slimit_clause ] [ soffset_clause ] [ timezone_clause ] .
```

#### Example

Select from measurements grouped by the day with a timezone

```sql
SELECT mean("value") FROM "cpu" GROUP BY region, time(1d) fill(0) tz('America/Chicago')
```

### SHOW CARDINALITY

Refers to the group of commands used to estimate or count exactly the cardinality of measurements, series, tag keys, tag key values, and field keys.

> The SHOW CARDINALITY commands are available in two variations: estimated and exact. Estimated values are calculated using sketches and are a safe default for all cardinality sizes. Exact values are counts directly from TSM (Time-Structured Merge Tree) data, but are expensive to run for high cardinality data.  Unless required, use the estimated variety.

Reviewer: ```suggestion
The SHOW CARDINALITY commands are available in two variations: _estimated_ and _exact_. Estimated values are calculated using sketches and are a safe default for all cardinality sizes. Exact values are counts directly from TSM (Time-Structured Merge Tree) data, but are expensive to run for high cardinality data.  Unless required, use the estimated variety.
```

-- show field keys and field value data types from specified measurement
SHOW FIELD KEYS FROM "cpu"
```

### SHOW MEASUREMENTS

```
show_measurements_stmt = "SHOW MEASUREMENTS" [on_clause] [ with_measurement_clause ] [ where_clause ] [ limit_clause ] [ offset_clause ] .
```

#### Examples

```sql
-- show all measurements
SHOW MEASUREMENTS

-- show measurements where region tag = 'uswest' AND host tag = 'serverA'
SHOW MEASUREMENTS WHERE "region" = 'uswest' AND "host" = 'serverA'

-- show measurements that start with 'h2o'
SHOW MEASUREMENTS WITH MEASUREMENT =~ /h2o.*/
```

### SHOW SERIES

```
show_series_stmt = "SHOW SERIES" [on_clause] [ from_clause ] [ where_clause ] [ limit_clause ] [ offset_clause ] .
```

#### Example

```sql
SHOW SERIES FROM "telegraf"."autogen"."cpu" WHERE cpu = 'cpu8'
```

### SHOW SERIES EXACT CARDINALITY

Estimates or counts exactly the cardinality of the series for the current database unless a database is specified using the ON <database> option.

#### Example

SHOW SERIES EXACT CARDINALITY" [ on_clause ] [ from_clause ] 
[ where_clause ] [ group_by_clause ] [ limit_clause ] [ offset_clause ]

```sql
SHOW SERIES EXACT CARDINALITY ON mydb
```

> - [Series cardinality](/influxdb/v2.7/reference/glossary/#series-cardinality) is the major factor that affects RAM requirements. For more information, see:
> 
> - [Don't have too many series](/influxdb/v2.7/reference/faq/#why-does-series-cardinality-matter).  As the number of unique series grows, so does the memory usage. High series cardinality can force the host operating system to kill the InfluxDB process with an out of memory (OOM) exception.

Reviewer: ```suggestion
```
